REST API Changes (#139)
lmstudio-ai/docs: 1_developer/0_core/_mcp.md

---
title: Using MCP via API
sidebar_title: Using MCP via API
description: Learn how to use Model Control Protocol (MCP) servers with LM Studio API.
index: 4
---

LM Studio supports Model Control Protocol (MCP) usage via API starting from version 0.4.0. MCP allows models to interact with external tools and services through standardized servers.

## How it works

MCP servers provide tools that models can call during chat requests. You can enable MCP servers in two ways: as ephemeral servers defined per-request, or as pre-configured servers in your `mcp.json` file.

## Ephemeral MCP servers

Ephemeral MCP servers are defined on-the-fly in each request. This is useful for testing or when you don't want to pre-configure servers.

```lms_info
Ephemeral MCP servers require the "Allow Ephemeral Remote MCPs" setting to be enabled in [Server Settings](/docs/developer/core/server/settings).
```

```lms_code_snippet
variants:
  curl:
    language: bash
    code: |
      curl http://localhost:1234/api/v1/chat \
        -H "Authorization: Bearer $LM_API_TOKEN" \
        -H "Content-Type: application/json" \
        -d '{
          "model": "qwen/qwen3-4b-2507",

>           "input": "What is the architecture of the Qwen3-Coder-480B-A35B-Instruct. Look at hugging face repo details",

Reviewer: ```suggestion
          "input": "What is the architecture of Qwen3-Coder-480B-A35B-Instruct as listed in the repo?",
```
But if there's a better example/model where the model doesn't need to be prompted so directly to use the tool, I think that would be better. Somewhat tough balance to strike between accessible model and interesting tool call.

Author: I have a task in the canvas for going through all examples due to showcasing the 4B example. I had to prompt it specifically, we can come back as a subsequent PR for better examples?

        "output": [
          {
            "type": "tool_call",
            "tool": "hub_repo_details",
            "arguments": {
              "repo_ids": [
                "Qwen/Qwen3-Coder-480B-A35B-Instruct"
              ],
              "repo_type": "model"
            },
            "output": "...",
            "provider_info": {
              "server_label": "huggingface",
              "type": "ephemeral_mcp"
            }
          },
          {
            "type": "message",
            "content": "The architecture of the **Qwen3-Coder-480B-A35B-Instruct** model is **qwen3_moe**. \n\nThis information is derived from the repository details provided by Hugging Face, which specify the model's architecture under the \"Technical Details\" section. Let me know if you'd like further details!"
          }
        ],
        "stats": {
          "input_tokens": 308,
          "total_output_tokens": 118,
          "reasoning_output_tokens": 0,
          "tokens_per_second": 49.44322495154522,
          "time_to_first_token_seconds": 0.217
        },
        "thread_id": "thread_0014d69c854c74594fb9bfef9ac4b3a6c8a776fad1f7c3d1"
      }
```

## MCP servers from mcp.json

> MCP servers can be pre-configured in your `mcp.json` file. This is the recommended approach for servers you use frequently.

Reviewer: ```suggestion
MCP servers can be pre-configured in your `mcp.json` file. This is the recommended approach for using MCP servers that take actions on your computer (like [microsoft/playwright-mcp](https://github.com/microsoft/playwright-mcp)) and servers that you use frequently.
```

          }
        ],
        "stats": {
          "input_tokens": 1428,
          "total_output_tokens": 460,
          "reasoning_output_tokens": 354,
          "tokens_per_second": 42.7891711368994,
          "time_to_first_token_seconds": 2.806
        },
        "thread_id": "thread_d26d96a8f3c0c15601f12fde1c682f24b65bb5b302c81795"
      }
```

## Ephemeral vs mcp.json servers

| Feature | Ephemeral | mcp.json |
|---------|-----------|--------|
| Configuration | Defined in each request | Pre-configured in `mcp.json` |
| Use case | Testing, one-off requests | Frequent use, stable servers |
| Server ID | Specified via `server_label` | Specified via `id` (e.g., `mcp/playwright`) |
| Custom headers | Supported via `headers` field | Configured in `mcp.json` |

## Restricting tool access

For both ephemeral and mcp.json servers, you can limit which tools the model can call using the `allowed_tools` field:

```lms_code_snippet
variants:
  curl:
    language: bash
    code: |
      curl http://localhost:1234/api/v1/chat \
        -H "Authorization: Bearer $LM_API_TOKEN" \
        -H "Content-Type: application/json" \
        -d '{
          "model": "qwen/qwen3-4b-2507",

>           "input": "Find trending models using HF MCP and open the top one in my browser. Only open the top one, not the rest",

Reviewer: Unfortunately again I think this is a bit too specific of prompting for a good example? Ideally the model does not have to be told exactly which tools to use in these examples

Reviewer: If we need to can change the model (ideally to something of similar size, or could switch to something bigger like `gpt-oss-20b` if needed)

MCP servers provide tools that models can call during chat requests. You can enable MCP servers in two ways: as ephemeral servers defined per-request, or as pre-configured servers in your `mcp.json` file.

> ## Ephemeral vs mcp.json servers

Reviewer: <img width="579" height="593" alt="Screenshot 2025-10-28 at 5 33 03 PM" src="https://github.com/user-attachments/assets/6cadfcb0-d2a5-46b3-af24-2bc1f3ce68cc" />

Oh no. May need html here too?
